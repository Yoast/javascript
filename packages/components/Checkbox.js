import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

const YoastCheckbox = styled.input`
	margin-right: 0.5em;
`;

/**
 * Creates a React Checkbox which consists of a checkbox and a label.
 */
class Checkbox extends React.Component {
	/**
	 * Constructs a checkbox component.
	 *
	 * @param {Object} props The props for this checkbox.
	 *
	 * @returns {void}
	 */
	constructor( props ) {
		super( props );

		this.state = {
			checked: props.checked,
		};
	}

	/**
	 * Handles changes in the Checkbox.
	 *
	 * @param {Event} event The onChange event.
	 * @returns {void} Fires the onChange function passed in the props.
	 */
	handleChange( event ) {
		this.setState( { checked: event.target.checked }, () => this.props.onChange( this.state.checked ) );
	}

	/**
	 * Renders a checkbox and a label.
	 *
	 * @returns {ReactElement} The Checkbox react component including its label.
	 */
	render() {
		return (
			<React.Fragment>
				<YoastCheckbox type="checkbox" id={ this.props.id } onChange={ this.handleChange.bind( this ) } />
				<label htmlFor={ this.props.id }>
					{ this.props.label }
				</label>
			</React.Fragment>
		);
	}
}

Checkbox.propTypes = {
	id: PropTypes.string.isRequired,
	onChange: PropTypes.func.isRequired,
	label: PropTypes.oneOfType( [
		PropTypes.string,
		PropTypes.array,
	] ).isRequired,
	checked: PropTypes.bool,

};

Checkbox.defaultProps = {
	checked: false,
};

export default Checkbox;
