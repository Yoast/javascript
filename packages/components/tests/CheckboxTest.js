import React from "react";
import renderer from "react-test-renderer";

import Checkbox from "../Checkbox";

describe( Checkbox, () => {
	it( "matches the snapshot", () => {
		const component = renderer.create(
			<Checkbox
				id="test-id" onChange={ () => {
				} } label="test label"
			/>
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );

	it( "matches the snapshot when an array is provided as a label", () => {
		const component = renderer.create(
			<Checkbox
				id="test-id" onChange={ () => {
				} } label={ [ "test label ", "using arrays" ] }
			/>
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );

	it( "executes callback once", () => {
		const event = {
			target: {
				checked: false,
			},
		};
		const onChange = jest.fn();
		const component = renderer.create(
			<Checkbox
				id="testCallback"
				onChange={ onChange }
				label="testCallbackLabel"
			/>
		);

		const tree = component.toJSON();
		tree[ 0 ].props.onChange( event );

		expect( onChange ).toHaveBeenCalledTimes( 1 );
	} );
} );
