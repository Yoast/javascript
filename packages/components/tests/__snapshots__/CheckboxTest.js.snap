// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox matches the snapshot 1`] = `
Array [
  <input
    className="Checkbox__YoastCheckbox-zj31xu-0 iwloWA"
    id="test-id"
    onChange={[Function]}
    type="checkbox"
  />,
  <label
    htmlFor="test-id"
  >
    test label
  </label>,
]
`;

exports[`Checkbox matches the snapshot when an array is provided as a label 1`] = `
Array [
  <input
    className="Checkbox__YoastCheckbox-zj31xu-0 iwloWA"
    id="test-id"
    onChange={[Function]}
    type="checkbox"
  />,
  <label
    htmlFor="test-id"
  >
    test label 
    using arrays
  </label>,
]
`;
